common/rconfiguration.cc

```diff
@@ -131,28 +131,92 @@ bool RWriteConfigFile(Configuration &Conf)
    return true;
 }
 
+static bool mkdirWithParents(const string &path)
+{
+   struct stat stbuf;
+   string delimeter("/");
+   const char *partial_path;
+   auto pos = path.find(delimeter, 1);
+
+   while (pos != string::npos) {
+      partial_path = path.substr(0, pos).c_str();
+
+      if ( stat(partial_path, &stbuf) < 0 &&
+	   mkdir(partial_path, 0700) < 0) {
+	 return _error->Errno("mkdir",
+			      _("ERROR: could not create directory %s"),
+			      partial_path);
+      }
+      pos = path.find(delimeter, pos + 1);
+   }
+   return true;
+}
 
 static bool checkConfigDir(string &path)
 {
    struct stat stbuf;
    struct passwd *pwd;
+   char *buf;
+   string xdg_data_dir;
+   string old_path;
+   string home_dir;
+   int l;
 
    pwd = getpwuid(getuid());
    if (!pwd) {
       return _error->Errno("getpwuid",
                            _
                            ("ERROR: Could not get password entry for superuser"));
    }
-   path = string(pwd->pw_dir) + "/.synaptic";
+
+   home_dir = string(pwd->pw_dir);
+   xdg_data_dir = home_dir + "/.config";
+   old_path = home_dir + "/.synaptic";
+   buf = getenv("XDG_CONFIG_HOME");
+
+   if (buf) {
+     l = strlen(buf) - 1;
+     if (buf[l] == '/') {
+       buf[l] = '\0';
+     }
+     xdg_data_dir=buf;
+
+     // Guarantee that XDG_CONFIG_HOME is inside home directory.
+     if (xdg_data_dir.substr(0,home_dir.length()) != home_dir) {
```

Comment:
(idle thinking) Same as above, slightly sad that there is no string::HasPrefix or similar. So this is fine too :)